## PROBLEM 1 - Education and church membership

<img src="SymPyBilleder/2025-11-27-09-41-48.png" width="750">

In [3]:
# We read the data by simply loading the exam2023.R file:
source("../data/exam2023.R")

church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)
summary(church_model)


Call:
lm(formula = church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

Residuals:
   Min     1Q Median     3Q    Max 
 -8546  -2284   -309   1807   7603 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 3224.8421   545.7549   5.909 5.62e-08 ***
H3020         22.5882     2.4702   9.144 1.32e-14 ***
H4058         19.0056     1.9203   9.897 3.38e-16 ***
H5020          3.6465     0.8627   4.227 5.53e-05 ***
H8090         33.5310     3.3995   9.863 3.98e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3072 on 93 degrees of freedom
Multiple R-squared:  0.9951,	Adjusted R-squared:  0.9949 
F-statistic:  4748 on 4 and 93 DF,  p-value: < 2.2e-16


### Question 1.1

<img src="SymPyBilleder/2025-11-27-09-42-31.png" width="750">

BOTH CHAT AND GEMINI HAD THIS ONE CORRECT

#### Chat

In [4]:
# fit model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# extract values
rse <- summary(church_model)$sigma        # residual standard error
df  <- church_model$df.residual           # residual degrees of freedom
SSR <- rse^2 * df                          # sum of squared residuals

y    <- exam2023$church
SST  <- sum((y - mean(y))^2)               # total sum of squares

R2_manual <- 1 - SSR / SST                 # manual R^2
R2_model  <- summary(church_model)$r.squared

cat("Manual R2:", R2_manual, "\n")
cat("Model R2 :", R2_model, "\n")

Manual R2: 0.995127 
Model R2 : 0.995127 


R² is

$$ R^2 = 1 - \frac{SSR}{SST} $$

where:

* $SSR$ is the sum of squared residuals
* $SST$ is the total sum of squares

From the output:

* Residual standard error $RSE = 3072$
* Degrees of freedom for residuals $df = 93$

**Step 1**

Compute the sum of squared residuals

$$ SSR = RSE^2 \cdot df $$

$$ SSR = 3072^2 \cdot 93 $$

**Step 2**

Compute the total sum of squares using the R² reported in the summary
R gives $R^2 = 0.9951$ so

$$ 0.9951 = 1 - \frac{SSR}{SST} $$

Rearrange

$$ \frac{SSR}{SST} = 1 - 0.9951 $$

$$ SST = \frac{SSR}{1 - 0.9951} $$

**Step 3**

Insert $SSR$ from Step 1 and calculate the ratio

$$ R^2 = 1 - \frac{SSR}{SST} $$

This yields

$$ R^2 = 0.9951 $$

No extra symbols appear after the equations and every formula is enclosed in $...$ or $$...$$ exactly as you requested.


#### Gemini

In [5]:
# Values from the summary output
F_stat <- 4748
df_model <- 4        # degrees of freedom (numerator)
df_residuals <- 93   # degrees of freedom (denominator)

# Calculate R-squared
# Formula: (F * df_model) / (F * df_model + df_residuals)
numerator <- F_stat * df_model
denominator <- numerator + df_residuals
R_squared <- numerator / denominator

# Print the result
R_squared

[1] 0.9951271

In [6]:
# Assuming your model is named 'church_model'
summary(church_model)$r.squared

[1] 0.995127

To solve this by hand, we use the relationship between the F-statistic and $R^2$. The formula for the F-statistic is defined as

$$F = \frac{\frac{R^2}{p}}{\frac{1 - R^2}{n - p - 1}}$$

where $p$ is the number of predictors and $n - p - 1$ is the degrees of freedom for the residuals. We can rearrange this formula to solve for $R^2$ directly

$$R^2 = \frac{F \cdot p}{F \cdot p + (n - p - 1)}$$

From the R summary output, we have the following values

$$F = 4748$$

$$p = 4$$

$$n - p - 1 = 93$$

Substituting these values into the rearranged equation gives

$$R^2 = \frac{4748 \cdot 4}{4748 \cdot 4 + 93}$$

$$R^2 = \frac{18992}{18992 + 93}$$

$$R^2 = \frac{18992}{19085}$$

$$R^2 \approx 0.995127$$

Rounding to four decimal places, we get

$$R^2 = 0.9951$$

### Question 1.2

<img src="SymPyBilleder/2025-11-27-11-03-42.png" width="750">

#### Chat

Chat got the code right, but not the hand calculation right

In [7]:
# fit model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# compute studentized residuals
rstud <- rstudent(church_model)

# count how many have |Rstudent| > 2
count_above2 <- sum(abs(rstud) > 2)

cat("Number of observations with |Rstudent| > 2:", count_above2, "\n")


Number of observations with |Rstudent| > 2: 5 


#### Gemini

Gemini got the code right, but didn't understand the hand calculation

In [8]:
# 1. Get the R-student residuals directly
r_vals <- rstudent(church_model)

# 2. Count how many are larger than 2 or smaller than -2
count <- sum(abs(r_vals) > 2)

# Print the result
count

[1] 5

### Question 1.3

<img src="SymPyBilleder/2025-11-27-11-06-28.png" width="750">

#### Chat

In [10]:
# fit model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# compute studentized residuals
rstud <- rstudent(church_model)

# compute leverages
lev <- hatvalues(church_model)

# logical vector: both conditions
infl <- abs(rstud) > 2 & lev > 0.102

# count them
count_influential <- sum(infl)

cat("Number of influential observations:", count_influential, "\n")

Number of influential observations: 2 


**Step 1**

Compute the studentized residuals

$$ R_i = \frac{e_i}{\hat{s}_{(i)} \sqrt{1 - h_i}} $$

**Step 2**

Compute the leverages

$$ h_i = x_i (X^\top X)^{-1} x_i^\top $$

**Step 3**

Identify the observations that satisfy

$$ |R_i| > 2 $$

and

$$ h_i > 0.102 $$

**Step 4**

Count how many indices satisfy both conditions

$$ 2 $$

This matches the R output and gives the correct answer: **2**


#### Gemini

In [9]:
# 1. Get the R-student residuals
r_vals <- rstudent(church_model)

# 2. Get the Leverage (hat values)
h_vals <- hatvalues(church_model)

# 3. Count observations that satisfy BOTH conditions simultaneously
#    Condition A: abs(r_vals) > 2
#    Condition B: h_vals > 0.102
count <- sum(abs(r_vals) > 2 & h_vals > 0.102)

# Print the result
count

[1] 2

### Question 1.4

<img src="SymPyBilleder/2025-11-27-11-20-03.png" width="750">

#### Chat

In [12]:
# fit the model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# Cook's distances
cooks <- cooks.distance(church_model)

# show first few if you like
head(cooks)

# which observation changes the parameter vector the most?
most_influential_index <- which.max(cooks)

cat("Most influential observation (by Cook's distance):",
    most_influential_index, "\n")

1            2            3            4            5            6 
1.086750e+01 5.866678e-02 2.093344e-05 1.109032e-05 1.590949e-02 1.521273e-03

Most influential observation (by Cook's distance): 1 


#### Gemini

In [11]:
# Calculate Cook's distance for every point
cooks_vals <- cooks.distance(church_model)

# Find which observation number has the maximum value
which.max(cooks_vals)

1 
1

### Question 1.5

<img src="SymPyBilleder/2025-11-27-11-22-41.png" width="750">

#### Chat

In [13]:
# fit the model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# if not installed, run once:
# install.packages("car")

library(car)

# variance inflation factors
vifs <- vif(church_model)

# tolerance values = 1 / VIF
tolerance <- 1 / vifs

result <- data.frame(
  Variable  = names(vifs),
  VIF       = as.numeric(vifs),
  Tolerance = as.numeric(tolerance)
)

print(result)

# quick check against the usual cutoffs
any_vif_big        <- any(vifs > 10)
any_tol_too_small  <- any(tolerance < 0.1)

cat("\nAny VIF > 10? ", any_vif_big, "\n")
cat("Any Tolerance < 0.1? ", any_tol_too_small, "\n")

Loading required package: carData

Warning message:
“package ‘carData’ was built under R version 4.5.1”


  Variable       VIF  Tolerance
1    H3020 26.833364 0.03726704
2    H4058 13.696907 0.07300918
3    H5020 56.060649 0.01783782
4    H8090  8.605281 0.11620772

Any VIF > 10?  TRUE 
Any Tolerance < 0.1?  TRUE 


Answer 1: Yes, there are several independent variables with a VIF larger than 10 and a Tolerance less than 0.1

In [15]:
# Example for variable H3020
# 1. Regress H3020 against all other predictors
m_H3020 <- lm(H3020 ~ H4058 + H5020 + H8090, data = exam2023)

# 2. Get the R-squared of this auxiliary model
r2_H3020 <- summary(m_H3020)$r.squared

# 3. Calculate VIF
vif_H3020 <- 1 / (1 - r2_H3020)

# 4. Calculate Tolerance
tol_H3020 <- 1 - r2_H3020

# Print results
vif_H3020
tol_H3020

# (Repeat these 4 steps for H4058, H5020, and H8090)

[1] 26.83336

[1] 0.03726704

### Question 1.6 

<img src="SymPyBilleder/2025-11-27-11-26-19.png" width="750">

#### Chat

In [16]:
# fit model
church_model <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# extract t-values
tvals <- summary(church_model)$coefficients[-1, 3]   # remove intercept

# find variable with smallest |t|
min_var <- names(which.min(abs(tvals)))
min_t   <- tvals[min_var]

# compute F = t^2
F_value <- min_t^2

cat("Variable removed first:", min_var, "\n")
cat("F-value for that variable:", F_value, "\n")

Variable removed first: H5020 
F-value for that variable: 17.86552 


#### Gemini

In [17]:
# Check which variable to drop and its F-value
drop1(church_model, test = "F")

Df Sum of Sq RSS        AIC      F value  Pr(>F)      
<none> NA        NA  877734461 1578.773       NA           NA
H3020   1 789182653 1666917114 1639.629 83.61753 1.322324e-14
H4058   1 924464750 1802199211 1647.276 97.95129 3.379127e-16
H5020   1 168614887 1046349347 1593.993 17.86552 5.534924e-05
H8090   1 918209483 1795943944 1646.935 97.28851 3.978427e-16

In [18]:
# Square the t-value of H5020 from the summary
t_val <- 4.227
f_val <- t_val^2
f_val

[1] 17.86753

### Question 1.6 

<img src="SymPyBilleder/2025-11-27-11-31-20.png" width="750">

#### Chat

In [19]:
vars <- c("H3020", "H4058", "H5020", "H8090")

f_values <- sapply(vars, function(v) {
  form <- as.formula(paste("church ~", v))
  model <- lm(form, data = exam2023)
  summary(model)$fstatistic[1]      # extract F-value
})

result <- data.frame(
  Variable = vars,
  F_value = as.numeric(f_values)
)

print(result)

best <- result$Variable[which.max(result$F_value)]

cat("\nFirst variable to enter in forward selection:", best, "\n")

  Variable   F_value
1    H3020 1022.4264
2    H4058 1934.2260
3    H5020 5533.1126
4    H8090  287.8724

First variable to enter in forward selection: H5020 


#### Gemini

In [21]:
# Calculate the correlation matrix
cor(exam2023[, c("church", "H3020", "H4058", "H5020", "H8090")])

# Or use the add1 command on an empty model to see the F-values explicitly
null_model <- lm(church ~ 1, data = exam2023)
add1(null_model, scope = ~ H3020 + H4058 + H5020 + H8090, test = "F")

church    H3020     H4058     H5020     H8090    
church 1.0000000 0.9561198 0.9760710 0.9914362 0.8659774
H3020  0.9561198 1.0000000 0.9287726 0.9559503 0.7065900
H4058  0.9760710 0.9287726 1.0000000 0.9619031 0.8217746
H5020  0.9914362 0.9559503 0.9619031 1.0000000 0.8572020
H8090  0.8659774 0.7065900 0.8217746 0.8572020 1.0000000

Df Sum of Sq    RSS          AIC      F value   Pr(>F)      
<none> NA           NA 180121413506 2092.529        NA           NA
H3020   1 164660712988  15460700518 1853.907 1022.4264 5.554046e-53
H4058   1 171604304732   8517108774 1795.477 1934.2260 2.027034e-65
H5020   1 177049587296   3071826210 1695.536 5533.1126 1.099598e-86
H8090   1 135076089840  45045323666 1958.705  287.8724 1.198811e-30

**Look for the variable with the highest F-value in the output.**

### Question 1.8 

<img src="SymPyBilleder/2025-11-27-11-33-56.png" width="750">

#### Chat

For this question you compare the **full model**

* M1

$$ \text{Church} = \alpha + \beta_1 H3020 + \beta_2 H4058 + \beta_3 H5020 + \beta_4 H8090 + \varepsilon $$

to the **reduced (simpler) model**

$$ \text{Church} = \tilde{\alpha} + \tilde{\beta}_1 H3020 + \tilde{\varepsilon} $$

You test

$$ H_0 : \beta_2 = \beta_3 = \beta_4 = 0 $$

The usual F–test statistic for comparing a reduced and a full model is

$$ F = \frac{(SSE_R - SSE_F) / (df_R - df_F)}{SSE_F / df_F} $$

where

* $SSE_R$ and $df_R$ are the residual sum of squares and residual df for the **reduced** model
* $SSE_F$ and $df_F$ are the residual sum of squares and residual df for the **full** model

In your case

* $SSE_R = 15460700518$
* $SSE_F = 877734461$
* $df_R = 96$
* $df_F = 93$

So the statistic is

$$ F = \frac{(15460700518 - 877734461) / (96 - 93)}{877734461 / 93} = 515 ....$$

This matches the **fourth** option in the list (the fraction with $(96-93)$ in the numerator’s denominator).

In [23]:
# full and reduced models
m_full    <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)
m_reduced <- lm(church ~ H3020, data = exam2023)

# sums of squares and dfs
SSE_full <- deviance(m_full)
SSE_red  <- deviance(m_reduced)

df_full <- df.residual(m_full)
df_red  <- df.residual(m_reduced)

# manual F statistic
F_stat <- ((SSE_red - SSE_full) / (df_red - df_full)) / (SSE_full / df_full)

cat("SSE_reduced:", SSE_red, "\n")
cat("SSE_full   :", SSE_full, "\n")
cat("df_reduced :", df_red,  "\n")
cat("df_full    :", df_full, "\n")
cat("F statistic:", F_stat,  "\n")

# check with built-in anova
anova(m_reduced, m_full)

SSE_reduced: 15460700518 
SSE_full   : 877734461 
df_reduced : 96 
df_full    : 93 
F statistic: 515.0441 


Res.Df RSS         Df Sum of Sq   F        Pr(>F)      
1 96     15460700518 NA          NA       NA           NA
2 93       877734461  3 14582966057 515.0441 8.799792e-58

#### Gemini

In [22]:
# 1. Fit the Full Model (M1)
model_1 <- lm(church ~ H3020 + H4058 + H5020 + H8090, data = exam2023)

# 2. Fit the Reduced Model (M2)
model_2 <- lm(church ~ H3020, data = exam2023)

# 3. Perform the F-test comparing them
anova(model_2, model_1)

Res.Df RSS         Df Sum of Sq   F        Pr(>F)      
1 96     15460700518 NA          NA       NA           NA
2 93       877734461  3 14582966057 515.0441 8.799792e-58

## PROBLEM 2 - Education and municipal compensation

<img src="SymPyBilleder/2025-11-27-11-40-35.png" width="750">

### Question 2.1

<img src="SymPyBilleder/2025-11-27-11-41-53.png" width="750">

#### Chat

In [25]:
# prepare variables
Xvars <- c("H70","H7020","H7025","H7030","H7035","H7039",
           "H7059","H7075","H7080","H7090","H7095")

# fit LDA with equal priors
library(MASS)
lda_model <- lda(gives ~ ., data = exam2023[, c("gives", Xvars)],
                 prior = c(0.5, 0.5))

# resubstitution prediction
pred <- predict(lda_model)$class

# count misclassifications
misclassified <- sum(pred != exam2023$gives)

cat("Number of resubstitution misclassifications:", misclassified, "\n")

Number of resubstitution misclassifications: 1 


#### Gemini

In [24]:
library(MASS)

# 1. Fit the LDA model
#    We include all variables listed in the problem description
model_lda <- lda(gives ~ H70 + H7020 + H7025 + H7030 + H7035 + H7039 + 
                         H7059 + H7075 + H7080 + H7090 + H7095, 
                 data = exam2023, 
                 prior = c(0.5, 0.5))

# 2. Predict classes for the same dataset (Resubstitution)
pred_classes <- predict(model_lda)$class

# 3. Create a confusion matrix (Predicted vs Actual)
conf_matrix <- table(Predicted = pred_classes, Actual = exam2023$gives)

# 4. Calculate total misclassifications
#    Sum of the off-diagonal elements (where predicted != actual)
misclassifications <- sum(pred_classes != exam2023$gives)

# Print the result
print(misclassifications)
print(conf_matrix)

[1] 1
         Actual
Predicted  G  R
        G  7  0
        R  1 90


### Question 2.2

<img src="SymPyBilleder/2025-11-27-11-48-11.png" width="750">

#### Gemini

In [26]:
# 1. Convert the class variable 'gives' to numeric (0 and 1)
#    (LDA for 2 classes is equivalent to Regression on the dummy variable)
exam2023$gives_num <- as.numeric(as.factor(exam2023$gives)) - 1

# 2. Fit the Full Model (All variables)
full_model <- lm(gives_num ~ H70 + H7020 + H7025 + H7030 + H7035 + H7039 + 
                 H7059 + H7075 + H7080 + H7090 + H7095, 
                 data = exam2023)

# 3. Fit the Reduced Model (Remove H7020, H7025, H7030)
reduced_model <- lm(gives_num ~ H70 + H7035 + H7039 + 
                    H7059 + H7075 + H7080 + H7090 + H7095, 
                    data = exam2023)

# 4. Perform the F-test (ANOVA)
anova(reduced_model, full_model)

Res.Df RSS      Df Sum of Sq F        Pr(>F)     
1 89     2.527747 NA        NA       NA          NA
2 86     2.123900  3 0.4038472 5.450799 0.001769614

### Question 2.3 

<img src="SymPyBilleder/2025-11-27-11-51-45.png" width="750">

#### Chat

In [29]:
# priors
pi_G <- 0.1

# squared generalized distance from group G mean to itself
D2 <- -2 * log(pi_G)

cat("Squared generalized distance for group G to itself:", D2, "\n")

Squared generalized distance for group G to itself: 4.60517 


#### Gemini

In [28]:
# Calculate the squared generalized distance penalty
# Formula: -2 * log(prior)
dist_G <- -2 * log(0.1)

# Print the result
dist_G

[1] 4.60517

## PROBLEM 3 - Education and correlation

<img src="SymPyBilleder/2025-11-27-12-07-16.png" width="750">

### Question 3.1

<img src="SymPyBilleder/2025-11-27-12-08-03.png" width="750">

#### Chat

In [32]:
# load exam data
source("../data/exam2023.R")

# extract variables
x <- exam2023$H7035
y <- exam2023$H7095

# correlation
r <- cor(x, y)

# fraction of variance explained
fraction <- r^2

# cat("Correlation:", r, "\n")
cat("Fraction of variance of H7035 explained by H7095:", fraction, "\n")

Fraction of variance of H7035 explained by H7095: 0.4855402 


#### Gemini

In [30]:
# 1. Calculate the correlation (r)
r_val <- cor(exam2023$H7035, exam2023$H7095)

# 2. Square it to get the explained variance (R^2)
result <- r_val^2

# Print the result
result

[1] 0.4855402

### Question 3.2

<img src="SymPyBilleder/2025-11-27-12-10-12.png" width="750">

#### Chat

In [35]:
# load data
source("../data/exam2023.R")

x <- exam2023$H7035
y <- exam2023$H7095

r <- cor(x, y)
n <- length(x)

t_stat <- r * sqrt((n - 2) / (1 - r^2))

# cat("Correlation r:", r, "\n")
cat("Test statistic t:", t_stat, "\n")

Test statistic t: 9.518588 


#### Gemini

In [33]:
# 1. Define the values (inferred from the problem context)
r <- 0.69681  # Correlation coefficient
n <- 98       # Number of observations (municipalities)

# 2. Calculate the t-statistic
#    Formula: r * sqrt(n - 2) / sqrt(1 - r^2)
t_stat <- r * sqrt(n - 2) / sqrt(1 - r^2)

# Print the result
t_stat

[1] 9.518663